blogposts/2020/performance-improvements-in-precise-code-intel.md

```diff
@@ -0,0 +1,256 @@
+
+After the port to Go, the API server became a pretty thin wrapper over queries to the bundle manager. Regardless of the performance outcome, this service was destined to be removed in favor of directly querying the bundle manager from the frontend (the only consumer of the API).
+
+In order to do this with a minimal diff, we replace the network calls to the api-server by [calling the HTTP handlers directly](https://github.com/sourcegraph/sourcegraph/pull/10872/files#diff-7c26d1b2bd39d2fa821872a65e774652R40) with a fake HTTP request. This allowed us to collapse the network boundary without touching much of the api-server client, or the HTTP handler code in the api-server. This is a pretty neat trick!
+
+Additional steps were required to reduce these layers into a direct function call, but the remaining steps were only to increase the hygiene of the code and were not huge performance boosters.
+
+## Increase parallelism
+
+This section describes a handful of changes that were made to split work so that parts of it can be done concurrently (switching to another pending task when the current task is blocked) or in parallel (performed at the same time on physical different cores).
+
+### Efficient input streaming
+
+**PRs**:
+- [codeintel: Refactor lsif input parsing (#10935)](https://github.com/sourcegraph/sourcegraph/pull/10935)
+- [codeintel: LSIF line reader (#10990)](https://github.com/sourcegraph/sourcegraph/pull/10990)
+
+Even after decreasing the CPU time and allocation pressure due to JSON parsing, it remained the bottleneck. The lines fed into the correlator process were consumed almost immediately, making this an issue of a slow-producer/fast-consumer (which is much more manageable problem than a slow consumer).
+
+In order to increase read throughput, we need to parallelize parsing of JSON lines while keeping the order the same (which matters to the correlator). Using channels as bounded queues allows us to break the problem into several parts, as illustrated below.
+
+![concurrency diagram](https://storage.googleapis.com/sourcegraph-assets/lsif-reader-parallelism.png)
+
+We continue to raw data line-by-line as fast as the input allows and as fast as the consumer can take input. Using a channel here acts as a read-ahead buffer. A number of unmarshaller routines read lines from the channel, parse it, and place the result onto an output channel. A batcher process consumes the items from the element channel in batches, reordering them by the input id (not order it was sent to the channel) so that the order is consistent with the raw LSIF input. After the batch receives the expected number of values from the channel, a signal is sent to the unmarshallers freeing them to resume work. This signaling procedure ensures that one unmarshaller is not looking for work past the the batching window. Each completed batch is free to be passed to the correlator.
```

Comment:
```suggestion
We continue to read data line-by-line as fast as the input allows and as fast as the consumer can take input. Using a channel here acts as a read-ahead buffer. A number of unmarshaller routines read lines from the channel, parse it, and place the result onto an output channel. A batcher process consumes the items from the element channel in batches, reordering them by the input id (not order it was sent to the channel) so that the order is consistent with the raw LSIF input. After the batch receives the expected number of values from the channel, a signal is sent to the unmarshallers freeing them to resume work. This signaling procedure ensures that one unmarshaller is not looking for work past the the batching window. Each completed batch is free to be passed to the correlator.
```

Reply:
Updated.